[@types/oauth2orize] Open interface

chdanielmueller opened the pull request:
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `pnpm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Adding an open interface like it has been done for Express.Request which allows to specify types for Client and User.
```typescript
declare global {
  namespace OAuth2orize {
    interface User extends Express.User {}
    interface Client {
      id: string;
      redirectUri: string;
    }
  }
}
```

Adding some more Arity to some functions

Infer Request Type on decision from MiddlewareFunction (can be expanded to additional functions)

DangerBotOSS commented:
<!--
  0 failure: 
  0 warning: 
  
  1 markdown notices
  DangerID: danger-id-Danger;
-->



## Formatting

The following files are not formatted:
1. types/oauth2orize/index.d.ts

Consider running `pnpm dprint fmt` on these files to make review easier.
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 5840f115bc31fbf58ae8a2872316f184526b6be6
</p>

weswigham reviewed (commented):
Are you sure this is a good approach? The `Express` declaration merging option is a legacy from before we had type parameter defaults - it's not a *great* solution, since it locks *every* client/user/whatever in the program into the same shape - which isn't necessarily true. Much better would be adding type parameters for `Client` and `User` to `createServer` and/or the `grant` methods that get flowed through to all the structures and signatures that use them - and these can be defaulted to `any` for back-compat. Alternatively - it seems like this is meant to pretty explicitly be used with `passport` and `express` - should it just be a direct reference to `Express.User` (or at least extend it)? And perhaps `Client` should be, too - at least in all the examples in their tests, the `user` and `client` always have the same shape.

chdanielmueller commented:
Hi @weswigham 

Thank you for your review.
I am not that "fluent" in more than basic Typescript Syntax and therefore very happy to receive feedback and ideas.

Passing Client and User Types to the createServer function seems like a great idea.
Do you have a sample where I can look at the syntax for the approach described by you?

The Express.User extension is not a good idea since the project can be used with other engines than express.
Client is something that is specific to oauth2orize and custom to the application/implementation. Therefore should also not be extended from somewhere else.

chdanielmueller commented:
Hi @weswigham 

I have rewritten the typings using parameter defaults and added some tests for the change.
Everything is backwards-compatible now as it defaults to any.

sheetalkamat commented on types/oauth2orize/.eslintrc.json:
```diff
@@ -1,5 +1,6 @@
     "rules": {
-        "@definitelytyped/no-any-union": "off"
+        "@definitelytyped/no-any-union": "off",
+        "@definitelytyped/no-unnecessary-generics": "off"
```
```suggestion
```

remove this and fix error or disable it on a line that needs to be.

chdanielmueller replied:
Changed by ab3d655213bead94dfb8f269182ed6366887979d
I disabled this rule as it errored at lots of places. It does not honor inheritance of types as a reason to keep it in.

sheetalkamat commented on types/oauth2orize/index.d.ts:
```diff
@@ -3,17 +3,9 @@
 
 import { IncomingMessage, ServerResponse } from "http";
 
-declare global {
-    namespace Express {
-        interface Request {
-            oauth2?: OAuth2 | undefined;
-        }
-    }
-}
-
```
i am not familiar with these types and someone more familiar with them needs to comment and see if this removal is ok. It seems not backward compatible.

chdanielmueller replied:
That's a fair point. I totally forgot that I removed this.
It has to be added by yourself in the project if you need the typing for the oauth2 property on Express Requests.
The issue is that I cannot override OAuth2<_any_, _any_> if it is already defined here.
I don't know if there is another solution to solve this.

chdanielmueller commented:
Ready to merge
